Kernel/Net/TCPSocket.cpp

```diff
@@ -664,9 +664,31 @@ ErrorOr<void> TCPSocket::close()
 {
     MutexLocker locker(mutex());
     auto result = IPv4Socket::close();
+    if (m_linger.l_onoff != 0) {
+        if (m_linger.l_linger > 0) {
+            dbgln_if(TCP_SOCKET_DEBUG, "SO_LINGER called with {} seconds timeout", m_linger.l_linger);
+            auto deadline = TimeManagement::the().current_time(CLOCK_MONOTONIC_COARSE) + Duration::from_seconds(m_linger.l_linger);
+            // FIXME: Return earlier if buffer empties before end of interval
+            (void)TimerQueue::the().add_timer_without_id(*m_timer, CLOCK_MONOTONIC_COARSE, deadline, [&]() {
+                auto has_unacked_data = m_unacked_packets.with_shared([&](auto const& packets) { return packets.size > 0; });
+                if (has_unacked_data) {
+                    dbgln_if(TCP_SOCKET_DEBUG, "SO_LINGER timeout expired. Some data couldn't make it :(");
+                    drop_receive_buffer();
+                    do_state_closed();
+                }
+                dbgln_if(TCP_SOCKET_DEBUG, "SO_LINGER timeout expired. No data left in buffer");
+                return;
+            });
+        } else {
+            dbgln_if(TCP_SOCKET_DEBUG, "SO_LINGER called without timeout, closing immediately");
+            TRY(send_tcp_packet(TCPFlags::RST));
+            do_state_closed();
+        }
+    }
+
     if (state() == State::CloseWait) {
         dbgln_if(TCP_SOCKET_DEBUG, " Sending FIN from CloseWait and moving into LastAck");
-        [[maybe_unused]] auto rc = send_tcp_packet(TCPFlags::FIN | TCPFlags::ACK);
+        TRY(send_tcp_packet(TCPFlags::FIN | TCPFlags::ACK));
```

Comment:
is the TRY() change right? we now don't do a bunch of teardown work on failing send that we did do before. In any case, it seems unrelated to the rest of this PR.